engine/src/main/java/com/datatorrent/stram/client/AppPackage.java

```diff
@@ -97,14 +98,33 @@ public AppPackage(File file) throws IOException, ZipException
    * If app directory is to be processed, there may be resource leak in the class loader. Only pass true for short-lived
    * applications
    *
+   * If contentFolder is set to a valid folder or the folder can be created, the file will be retained on disk after App Package is closed
+   * otherwise temp folder will be created and will be cleaned on close()
+   *
    * @param file
+   * @param contentFolder  the folder that the app package will be extracted to
    * @param processAppDirectory
    * @throws java.io.IOException
    * @throws net.lingala.zip4j.exception.ZipException
    */
-  public AppPackage(File file, boolean processAppDirectory) throws IOException, ZipException
+  public AppPackage(File file, File contentFolder, boolean processAppDirectory) throws IOException, ZipException
   {
     super(file);
+
+    boolean isContentFolderValid = false;
+    if (contentFolder != null) {
+      try {
+        FileUtils.forceMkdir(contentFolder);
+        isContentFolderValid = true;
+      } catch (Exception e) {
+        LOG.warn("Error creating writable app package directory {}, create temp folder instead", contentFolder.getAbsolutePath(), e);
+      }
+    }
+    if (!isContentFolderValid) {
```

Comment:
If the content folder given is not null but is invalid, I think we should throw an exception instead and not create a temp directory quietly.  The caller should know that the content folder is invalid so that the caller does not expect to find information in the invalid content folder.